ZIP 316: MUST-understand Metadata Items and Address Expiration Metadata (#759)
zcash/zips: zip-0316.rst

Terminology
===========

> The key words "MUST", "MUST NOT", "SHOULD", "MAY", and "RECOMMENDED" in this

Reviewer: ```suggestion
The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED", and "MAY" in this
```
(Because SHOULD and RECOMMENDED mean the same thing.)

Concepts
--------

> 

Reviewer: This change is unnecessary.

Privacy impacts of transparent or cross-pool transactions, and the
associated UX issues, will be addressed in ZIP 315 (in preparation).

Revision History
================

.. _`Revision 1`:

> * Revision 1: `MUST-understand typecodes`_ and `Address Expiration Metadata`_

Reviewer: ```suggestion
* Revision 1: `MUST-understand Typecodes`_ and `Address Expiration Metadata`_
```

Viewing Keys. These Items MAY affect the overall interpretation of
the UA / UVK (for example, by specifying an expiration date).

.. _`MUST-understand typecodes`:

> As of `Revision 1`_, the subset of Metadata typecodes in
> the range :math:`mathtt{0xE0}` to :math:`mathtt{0xEF}` are designated

Reviewer: ```suggestion
As of `Revision 1`_, the subset of Metadata typecodes in the range
:math:`mathtt{0xE0}` to :math:`mathtt{0xEF}` inclusive are designated
```

> as "MUST-understand": if a Consumer is unable to recognize the meaning

Reviewer: Consider changing "MUST-understand" to "MUST-recognize" everywhere. ("Understand" is a less accurate metaphor than "recognize".)

Author: Hm, I'm not sure about that. Take for example this sentence:
> As of `Revision 1`_, a Consumer of a UIVK MUST understand the meaning of all "MUST-recognise" Metadata Item typecodes present in the UIVK in order to perform Unified Address derivation, and consumers MUST follow the derivation directives specified for each such Metadata Item.

It's not just that the consumer must be able to identify the typecode, its meaning and the constraints it imposes must also be understood.

The value of a :math:`\mathtt{0xE0}` item MUST be an unsigned 32-bit integer in
little-endian order specifying the Address Expiry Height, a block height of the
Zcash chain associated with the UA/UVK. A Unified Address containing metadata
typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
the Zcash chain is greater than this value.

The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
little-endian order specifying a Unix Epoch Time, hereafter referred to as the
Address Expiry Time. A Unified Address containing metadata typecode
:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
after the Address Expiry Time.

A Sender that supports `Revision 1`_ of this specification MUST set
a non-zero ``nExpiryHeight`` field in transactions that it creates. If the
``nExpiryHeight`` normally constructed by the Sender would be greater than the
Address Expiry Height, then the transaction MUST NOT be sent. If only an
Address Expiry Time is specified, then the Sender MUST choose a value for
``nExpiryHeight`` such that the transaction will expire no more than 24 hours
after the current time.

> 

Reviewer: Specify that if a wallet user attempts to send to an expired address, the error presented to the user SHOULD include a suggestion to attempt to find a more recent address for the intended recipient.

When deriving from a UFVK or UIVK containing typecodes :math:`\mathtt{0xE0}`
and/or :math:`\mathtt{0xE1}`, these Metadata Items MUST be retained unmodified
in the output.

The intent of this specification is that Consumers of Unified Addresses MUST
NOT send to expired addresses. If only an Address Expiry Time is specified, a
transaction to the associated address could be mined after the Address Expiry
Time within a 24-hour window.

> 

Reviewer: Add a Rationale subsection explaining:
* That the reason to specify that the transaction MUST NOT be sent when its `nExpiryHeight` as normally constructed is greater than the Address Expiry Height &mdash;rather than limiting the `nExpiryHeight` field&mdash;is to avoid unnecessary information leakage in that field about which address was used as the destination.
* The intent behind requiring the Sender to choose a transaction that expires no more than 24 hours after the current time (i.e. that an adversary could reintroduce the tx an arbitrary time later, if it is still valid).

Also say that if a wallet user attempts to send to an expired address, the error presented to the user SHOULD include a suggestion to attempt to find a more recent address for the intended recipient.

a non-zero ``nExpiryHeight`` field in transactions that it creates. If the
``nExpiryHeight`` normally constructed by the Sender would be greater than the
Address Expiry Height, then the transaction MUST NOT be sent. If only an
Address Expiry Time is specified, then the Sender MUST choose a value for
``nExpiryHeight`` such that the transaction will expire no more than 24 hours
after the current time.

If wallet user attempts to send to an expired address, the error presented to
the user by the wallet SHOULD include a suggestion that the user should
attempt to obtain a currently-valid address for the intended recipient.

Address expiration imposes no constraints on the Producer of an address. A
Producer MAY generate multiple Unified Addresses with the same Receivers but
different expiration metadata and/or any number of distinct Diversified Unified
Addresses with the same or different expiry metadata, in any combination.

When deriving a UIVK from a UFVK containing Typecodes :math:`\mathtt{0xE0}`
and/or :math:`\mathtt{0xE1}`, these Metadata Items MUST be retained unmodified
in the derived UIVK.

When deriving a Unified Address from a UFVK or UIVK containing a Metadata Item

> having Typecode :math:`\mathtt{0xE0}` the derived Unified Address MUST contain

Reviewer: ```suggestion
having Typecode :math:`\mathtt{0xE0}`, the derived Unified Address MUST contain
```

a Metadata Item having Typecode :math:`\mathtt{0xE0}` such that the Address
Expiry Height of the resulting address is less than or equal to the Expiry Height
of the viewing key. 

When deriving a Unified Address from a UFVK or UIVK containing a Metadata Item

> having Typecode :math:`\mathtt{0xE1}` the derived Unified Address MUST contain

Reviewer: ```suggestion
having Typecode :math:`\mathtt{0xE1}`, the derived Unified Address MUST contain
```

a Metadata Item having Typecode :math:`\mathtt{0xE1}` such that the Address
Expiry Time of the resulting address is less than or equal to the Expiry Time
of the viewing key.

Producers of Diversified Unified Addresses should be aware that the expiration
metadata could potentially be used to link addresses from the same source.
Normally, if Diversified Unified Addresses derived from the same UIVK contain
only Sapling and/or Orchard Receivers and no Metadata Items, they will be
unlinkable as described in [#protocol-concretediversifyhash]_; this property

> does not hold when Metadata Items are present. It is RECOMMENDED that when

Reviewer: Add RECOMMENDED to the BCP 14 boilerplate.

Reviewer: Btw, this could be interpreted this in a silly way such as incrementing the expiry time by one second for each Diversified Unified Address. I guess we have to hope that implementors will think about it more than that.

deriving Unified Addresses from a UFVK or UIVK containing expiry metadata that
the Expiry Height and Expiry Time of each distinct address vary from one
another, so as to reduce the likelihood that addresses may be linked via their

> expiry metadata.

Reviewer: Btw, this could be interpreted this in a silly way such as incrementing the expiry time by one second for each Diversified Unified Address. I guess we have to hope that implementors will think about it more than that.

are reserved for optional address expiry metadata. A producer MAY choose to
generate Unified Addresses containing either or both of the following Metadata
Item Types, or none.

The value of a :math:`\mathtt{0xE0}` item MUST be an unsigned 32-bit integer in
little-endian order specifying the Address Expiry Height, a block height of the
Zcash chain associated with the UA/UVK. A Unified Address containing metadata
Typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
the Zcash chain is greater than this value.

The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
little-endian order specifying a Unix Epoch Time, hereafter referred to as the
Address Expiry Time. A Unified Address containing Metadata Typecode
:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
after the Address Expiry Time.

A Sender that supports `Revision 1`_ of this specification MUST set
a non-zero ``nExpiryHeight`` field in transactions that it creates. If the
``nExpiryHeight`` normally constructed by the Sender would be greater than the
Address Expiry Height, then the transaction MUST NOT be sent. If only an

> Address Expiry Time is specified, then the Sender MUST choose a value for

Reviewer: ```suggestion
Address Expiry Time is specified, then the Sender SHOULD choose a value for
```
Also explain in the rationale subsection that the reason for this being SHOULD rather than MUST, is just that MUST doesn't make sense for requirements involving local time.

``nExpiryHeight`` such that the transaction will expire no more than 24 hours
after the current time.

> 

Reviewer: Clarify that if both an 0xE0 and an 0xE1 Metadata Item are present, that both apply.

Reviewer: I think we lost the requirement to not send to an expired address at some point during editing (maybe when we made the change to say "the ``nExpiryHeight`` normally constructed by the Sender"?)

of the viewing key. 

When deriving a Unified Address from a UFVK or UIVK containing a Metadata Item
having Typecode :math:`\mathtt{0xE1}`, the derived Unified Address MUST contain
a Metadata Item having Typecode :math:`\mathtt{0xE1}` such that the Address
Expiry Time of the resulting address is less than or equal to the Expiry Time
of the viewing key.

Producers of Diversified Unified Addresses should be aware that the expiration
metadata could potentially be used to link addresses from the same source.
Normally, if Diversified Unified Addresses derived from the same UIVK contain
only Sapling and/or Orchard Receivers and no Metadata Items, they will be
unlinkable as described in [#protocol-concretediversifyhash]_; this property
does not hold when Metadata Items are present. It is RECOMMENDED that when
deriving Unified Addresses from a UFVK or UIVK containing expiry metadata that
the Expiry Height and Expiry Time of each distinct address vary from one
another, so as to reduce the likelihood that addresses may be linked via their
expiry metadata.

Rationale
'''''''''

The intent of this specification is that Consumers of Unified Addresses MUST

> NOT send to expired addresses. If only an Address Expiry Time is specified, a

Reviewer: This is in a Rationale subsection so is not spec.

Viewing Keys. These Items MAY affect the overall interpretation of
the UA / UVK (for example, by specifying an expiration date).

> .. _`MUST-understand typecodes`:

Reviewer: ```suggestion
.. _`MUST-understand Typecodes`:
```

> As of `Revision 1`_, the subset of Metadata typecodes in the range

Reviewer: ```suggestion
As of `Revision 1`_, the subset of Metadata Typecodes in the range
```

:math:`mathtt{0xE0}` to :math:`mathtt{0xEF}` inclusive are designated
as "MUST-understand": if a Consumer is unable to recognise the meaning

> of a Metadata Item with a typecode in this range, then it MUST regard

Reviewer: ```suggestion
of a Metadata Item with a Typecode in this range, then it MUST regard
```

the entire UA/UVK as unsupported and not process it further.

Since Metadata Items are not Receivers, they MUST NOT be selected by
a Sender when choosing a Receiver to send to, and since they are not
Viewing Keys, they MUST NOT provide additional authority to view
information about transactions.

Address Expiration Metadata
---------------------------

> As of `Revision 1`_, typecodes :math:`\mathtt{0xE0}` and :math:`\mathtt{0xE1}`

Reviewer: ```suggestion
As of `Revision 1`_, Typecodes :math:`\mathtt{0xE0}` and :math:`\mathtt{0xE1}`
```

are reserved for optional address expiry metadata. A producer MAY choose to
generate Unified Addresses containing either or both of the following Metadata
Item Types, or none.

The value of a :math:`\mathtt{0xE0}` item MUST be an unsigned 32-bit integer in
little-endian order specifying the Address Expiry Height, a block height of the
Zcash chain associated with the UA/UVK. A Unified Address containing metadata

> typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of

Reviewer: ```suggestion
Typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
```

the Zcash chain is greater than this value.

The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
little-endian order specifying a Unix Epoch Time, hereafter referred to as the

> Address Expiry Time. A Unified Address containing metadata typecode

Reviewer: ```suggestion
Address Expiry Time. A Unified Address containing Metadata Typecode
```

:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
after the Address Expiry Time.

A Sender that supports `Revision 1`_ of this specification MUST set
a non-zero ``nExpiryHeight`` field in transactions that it creates. If the
``nExpiryHeight`` normally constructed by the Sender would be greater than the
Address Expiry Height, then the transaction MUST NOT be sent. If only an
Address Expiry Time is specified, then the Sender MUST choose a value for
``nExpiryHeight`` such that the transaction will expire no more than 24 hours
after the current time.

If wallet user attempts to send to an expired address, the error presented to
the user by the wallet SHOULD include a suggestion that the user should
attempt to obtain a currently-valid address for the intended recipient.

> When deriving from a UFVK or UIVK containing typecodes :math:`\mathtt{0xE0}`

Reviewer: ```suggestion
When deriving from a UFVK or UIVK containing Typecodes :math:`\mathtt{0xE0}`
```

and/or :math:`\mathtt{0xE1}`, these Metadata Items MUST be retained unmodified
in the output.

Address expiration imposes no constraints on the Producer of an address. A
Producer MAY generate multiple Unified Addresses with the same Receivers but
different expiration metadata and/or any number of distinct Diversified Unified
Addresses with the same or different expiry metadata, in any combination.

> 

Reviewer: ```suggestion

Producers of Diversified Unified Addresses should be aware that the expiration
metadata could potentially be used to link addresses from the same source.
Normally, if Diversified Unified Addresses derived from the same UIVK contain
only Sapling and/or Orchard Receivers and no Metadata Items, they will be
unlinkable as described in [#protocol-concretediversifyhash]_; this property
does not hold when Metadata Items are present.

```
Also add this reference:
```rst
.. [#protocol-concretediversifyhash] `Zcash Protocol Specification, Version 2023.4.0. Section 5.4.1.6: DiversifyHash^Sapling and DiversifyHash^Orchard Hash Functions <protocol/protocol.pdf#concretediversifyhash>`_
```
(Optionally update the version for the other protocol spec references to 2023.4.0 as well. I have checked that none of the referenced sections were renumbered.)

Author: What we should do instead is, for address derivation, specify that the UIVK expiry height and time MUST be treated as upper bounds, and recommend that expiry heights and times be selected in a fashion that avoids this sort of linkability.

Rationale
'''''''''

The intent of this specification is that Consumers of Unified Addresses MUST
NOT send to expired addresses. If only an Address Expiry Time is specified, a
transaction to the associated address could be mined after the Address Expiry
Time within a 24-hour window.

> The reason that the transaction MUST NOT be sent when its nExpiryHeight as

Reviewer: ```suggestion
The reason that the transaction MUST NOT be sent when its ``nExpiryHeight`` as
```

normally constructed is greater than the Address Expiry Height is to avoid
unnecessary information leakage in that field about which address was used as
the destination. If a sender were to instead use the expiry height to directly

> set nExpiryHeight field, this would leak the expiry information of the

Reviewer: ```suggestion
set the ``nExpiryHeight`` field, this would leak the expiry information of the
```

destination address, which may then be identifiable.

When honoring an Address Expiry Time, the reason that a sender must choose a
``nExpiryHeight`` that is expected to occur within 24 hours of the time of
transaction construction is to ensure that the expiry time is respected to
within a day; since block times are not exact and computer clocks often
disagree, an Address Expiry Time is advisory and not a hard bound. Setting an
expiry height as specified ensures that the expiry time will be at least
loosely respected, and it is of course both permissible and perhaps advisable
to set this bound more tightly. A common expiry delta used by many wallets

> is 40 blocks from the current chain tip.

Reviewer: ```suggestion
is 40 blocks from the current chain tip, as suggested in
ZIP 203 [#zip-0203-default-expiry]_.
```

Deriving a UIVK from a UFVK
---------------------------

As of `Revision 1`_, a Consumer of a UIVK MUST understand the
meaning of all "MUST-understand" Metadata Item typecodes present in the

> UFVK in order to perform UIVK derivation, and consumers MUST follow the
> derivation directives specified for each such Metadata Item.

Reviewer: ```suggestion
UFVK in order to perform UIVK derivation. For Metadata Items recognised
by the Consumer, the processing of the Item when deriving a UIVK is
specified in the section or ZIP describing that Item.
```
The difference here is that we're saying that such processing might be specified (and where to look for the specification) for all Metadata Items, not just MUST-understand ones. It's not necessary to use MUST here: the relevant section or ZIP will say what the conformance requirement is, and it might not be a MUST in general (but it is for 0xE0 and 0xE1).

In each case, the Typecode remains the same as in the FVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognized by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be

> dropped when deriving a UIVK from a UFVK.

Reviewer: ```suggestion
dropped when deriving a UIVK from a UFVK. Metadata Items that are
recognized SHOULD be processed as specified in the description of those
Items.
```

Deriving a Unified Address from a UIVK
--------------------------------------

> As of `Revision 1`_, a Consumer of a UIVK MUST understand
> the meaning of all "MUST-understand" Metadata Item typecodes present in
> the UIVK in order to perform Unified Address derivation, and consumers
> MUST follow the derivation directives specified for each such Metadata
> Item.

Reviewer: ```suggestion
As of `Revision 1`_, a Consumer of a UIVK MUST understand the meaning of
all "MUST-understand" Metadata Item typecodes present in the UIVK in order
to perform Unified Address derivation. For Metadata Items recognised by
the Consumer, the processing of the Item when deriving a UIVK is specified
in the section or ZIP describing that Item.
```

Deriving a UIVK from a UFVK
---------------------------

As of `Revision 1`_, a Consumer of a UIVK MUST understand the

> meaning of all "MUST-understand" Metadata Item typecodes present in the

Reviewer: ```suggestion
meaning of all "MUST-understand" Metadata Item Typecodes present in the
```

In each case, the Typecode remains the same as in the FVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognised by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be
dropped when deriving a UIVK from a UFVK.

Deriving a Unified Address from a UIVK
--------------------------------------

As of `Revision 1`_, a Consumer of a UIVK MUST understand the meaning

> of all "MUST-understand" Metadata Item typecodes present in the UIVK

Reviewer: ```suggestion
of all "MUST-understand" Metadata Item Typecodes present in the UIVK
```

Deriving a UIVK from a UFVK
---------------------------

As of `Revision 1`_, a Consumer of a UIVK MUST understand the
meaning of all "MUST-understand" Metadata Item Typecodes present in the

> UFVK in order to perform UIVK derivation. For Metadata Items recognised

Reviewer: Express this as a consequence of the general MUST-understand requirement rather than restating it as if it were a separate conformance requirement.

Deriving a UIVK from a UFVK
---------------------------

> As of `Revision 1`_, a Consumer of a UIVK MUST understand the
> meaning of all "MUST-understand" Metadata Item Typecodes present in the
> UFVK in order to perform UIVK derivation. For Metadata Items recognised
> by the Consumer, the processing of the Item when deriving a UIVK is
> specified in the section or ZIP describing that Item.

Reviewer: Express this as a consequence of the general MUST-understand requirement:

```suggestion
As a consequence of the specification in `MUST-understand Typecodes`_,
as of `Revision 1`_ a Consumer of a UIVK MUST understand the meaning of
all "MUST-understand" Metadata Item Typecodes present in the UFVK in
order to perform UIVK derivation.

For Metadata Items recognised by the Consumer, the processing of the
Item when deriving a UIVK is specified in the section or ZIP describing
that Item.
```

In each case, the Typecode remains the same as in the IVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognized by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be

> dropped when deriving a Unified Address from a UIVK.

Reviewer: ```suggestion
dropped when deriving a Unified Address from a UIVK. Metadata Items that are
recognized SHOULD be processed as specified in the description of those
Items.
```

In each case, the Typecode remains the same as in the FVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognised by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be
dropped when deriving a UIVK from a UFVK.

Deriving a Unified Address from a UIVK
--------------------------------------

As of `Revision 1`_, a Consumer of a UIVK MUST understand the meaning
of all "MUST-understand" Metadata Item Typecodes present in the UIVK

> in order to perform Unified Address derivation. For Metadata Items

Reviewer: Express this as a consequence of the general MUST-understand requirement rather than restating it as if it were a separate conformance requirement.

recognised by the Consumer, the processing of the Item when deriving

> a UIVK is specified in the section or ZIP describing that Item.

Reviewer: ```suggestion
a Unified Address is specified in the section or ZIP describing that Item.
```
(copy-and-paste error from above)

In each case, the Typecode remains the same as in the FVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognised by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be
dropped when deriving a UIVK from a UFVK.

Deriving a Unified Address from a UIVK
--------------------------------------

> As of `Revision 1`_, a Consumer of a UIVK MUST understand the meaning
> of all "MUST-understand" Metadata Item Typecodes present in the UIVK
> in order to perform Unified Address derivation. For Metadata Items
> recognised by the Consumer, the processing of the Item when deriving
> a UIVK is specified in the section or ZIP describing that Item.

Reviewer: Express this as a consequence of the general MUST-understand requirement:
```suggestion
As of `Revision 1`_, a Consumer of a UIVK MUST understand the meaning
of all "MUST-understand" Metadata Item Typecodes present in the UIVK
in order to perform Unified Address derivation. For Metadata Items
recognised by the Consumer, the processing of the Item when deriving
a UIVK is specified in the section or ZIP describing that Item.
```

In each case, the Typecode remains the same as in the IVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognised by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be
dropped when deriving a Unified Address from a UIVK.

> 

Reviewer: ```suggestion

See `Address Expiration Metadata`_ for discussion of potential linking of
Diversified Unified Addresses via their metadata.

```

==========

.. [#BCP14] `Information on BCP 14 — "RFC 2119: Key words for use in RFCs to Indicate Requirement Levels" and "RFC 8174: Ambiguity of Uppercase vs Lowercase in RFC 2119 Key Words" <https://www.rfc-editor.org/info/bcp14>`_

> .. [#protocol-nu5] `Zcash Protocol Specification, Version 2023.4.0 or later [NU5 proposal] <protocol/protocol.pdf>`_

Reviewer: ```suggestion
.. [#protocol-nu5] `Zcash Protocol Specification, Version 2023.4.0 or later <protocol/protocol.pdf>`_
```
(Version 2023.4.0 no longer has "[NU5 proposal]" in its title. Non-blocking: the reference anchor could be renamed to just `[#protocol]`.)
